### ISSUE TITLE
Support ACL based on Domain/URL instead of IP

### ISSUE THREAD

**Issue by btzq** [apache/cloudstack#8917]
##### ISSUE TYPE

 * Improvement Request

##### COMPONENT NAME

~~~
Improvement Request (Functionality)
~~~

##### CLOUDSTACK VERSION

~~~
4.19
~~~

##### CONFIGURATION

N/A

##### OS / ENVIRONMENT

N/A

##### SUMMARY

Note: This ticket is broken down as requested from the origianl post, https://github.com/apache/cloudstack/issues/8841

**Support ACL based on Domain/URL instead of IP**

- As a company, sometimes i would like to block access to specific sites (eg. google drive), which is based on domain. 
- But i am unable to do so because, ACL Rules are based on explicitly specifying the IP address.
- Specifying the IP address/range of the intended website will be tedious and non-practical.

##### STEPS TO REPRODUCE

N/A

~~~
N/A
~~~

##### EXPECTED RESULTS

~~~
To be able to manage ACL Rules more easily for enterprise use cases.
~~~

##### ACTUAL RESULTS

~~~
Unable to effectively block access to certain sites using URL/Domain
~~~

**Comment by DaanHoogland** [comment]
As ACLs are implemented using iptables there is not a trivial way to implement this. The VR will have to use ipset and dig/nslookup to create sets of ips from domain names and (re)apply those. It will be hard to make this relyable if users try to change those on the fly, or reorder them. There also needs to be a mechanism to update them automatically as DNS entruies for them may change out of bounds.

**Comment by btzq** [comment]
We got this idea on some enterprise firewalls like PFSense.

Is there no way to achieve a similar result?

**Comment by DaanHoogland** [comment]
Yes @btzq , there is, as described above, but the programming needs to be done on the VR. I am just saying it is not a trivial job. It is certainly possible.

**Comment by vishesh92** [comment]
The workaround is to fetch the get the IP ranges of the providers you want to block (google, onedrive, etc.) and update the ACLs. This will need to be run as a script which runs frequently to ensure new and old IPs are added and removed to avoid blocking the actual traffic.
